File: Java/src/Shape/Shape3D/RotationSolid/Cylinder.java
package Shape.Shape3D.RotationSolid;

import Color.ColorRGB;
import Shape.Point2D;
import Shape.Point3D;
import Shape.Shape2D.Polygon.Quadrilateral.Rectangle;
import Shape.Shape3D.Shape3D;

import java.util.ArrayList;

public class Cylinder extends RotationSolid {

    public Cylinder(ArrayList<Point3D> p, Rectangle bShape) {
        super(p, bShape, "Cylinder");
    }

    public Cylinder(ArrayList<Point3D> p, Rectangle bShape, ColorRGB color) {
        super(p, bShape, "Cylinder", color);
    }

    public Cylinder(Rectangle r) {
        super(null, r, "Cylinder");
        points = buildCylinder(r);
    }

    public Cylinder(Rectangle r, ColorRGB color) {
        super(null, r, "Cylinder", color);
        points = buildCylinder(r);
    }

    public Cylinder(Cylinder c) {
        super(c);
    }

    private ArrayList<Point3D> buildCylinder(Rectangle r) {
        ArrayList<Point3D> points = new ArrayList<>();
        for (Point2D p2d : r.getPoints()) {
            points.add(Point3D.p2dTop3d(p2d));
        }
        return points;
    }

    public Cylinder sum(Shape3D other) {
        Rectangle newbaseShape = (Rectangle) baseShape.sum(other.getBaseShape());
        return new Cylinder(
                buildCylinder(newbaseShape), newbaseShape, getColor().sum(other.getColor()));
    }

    public Cylinder sub(Shape3D other) {
        Rectangle newbaseShape = (Rectangle) baseShape.sub(other.getBaseShape());
        return new Cylinder(
                buildCylinder(newbaseShape), newbaseShape, getColor().sub(other.getColor()));
    }

    public Cylinder mul(Shape3D other) {
        Rectangle newbaseShape = (Rectangle) baseShape.mul(other.getBaseShape());
        return new Cylinder(
                buildCylinder(newbaseShape), newbaseShape, getColor().mul(other.getColor()));
    }

    public Cylinder div(Shape3D other) {
        Rectangle newbaseShape = (Rectangle) baseShape.div(other.getBaseShape());
        return new Cylinder(
                buildCylinder(newbaseShape), newbaseShape, getColor().div(other.getColor()));
    }

    public double height() {
        return ((Rectangle) baseShape).height();
    }

    public double ray() {
        return ((Rectangle) baseShape).base();
    }

    public double lateralSurface() {
        Rectangle r = (Rectangle) baseShape;
        return 2 * Math.PI * r.base() * r.height();
    }

    public double baseSurface() {
        Rectangle r = (Rectangle) baseShape;
        return 2 * Math.PI * r.base() * r.height();
    }

    public double volume() {
        return Math.PI * Math.pow(ray(), 2) * height();
    }

    public double totalSurface() {
        return lateralSurface() + 2 * Math.PI * Math.pow(ray(), 2);
    }

    public String toString() {
        return super.toString() + "BaseRay " + ray() + "\nHeight " + height() + "\nVolume " + volume() +
                "\nLateralSurface " + lateralSurface() + "\nTotalSurface " + totalSurface();
    }
}

File: cpp/KalkCpp/Shape/Shape3D/RotationSolid/cylinder.cpp
#include "cylinder.h"
#include "../..//Shape2D/Polygon/Quadrilateral/rectangle.h"
#include "math.h"



Cylinder::Cylinder(const std::vector<Point3D>& vp3d, Rectangle* bs, ColorRGB* color):
    RotationSolid(vp3d,bs, "Cylinder", color)
{

}

Cylinder* Cylinder::clone() const
{
    return new Cylinder(*this);
}

std::vector<Point3D> Cylinder::buildCylinder(Rectangle* p) const
{
    //if is rectangle
    std::vector<Point3D> points = std::vector<Point3D>();
    for(std::vector<Point2D>::const_iterator it = p->getPoints().begin(); it != p->getPoints().end(); ++it)
    {
        points.push_back(Point3D::p2dTop3d(*it));
    }

    return points; //solo quelli 2d, dovrebbero bastare
}



Cylinder::Cylinder(Rectangle* bs, ColorRGB* color):
    RotationSolid(buildCylinder(bs), bs, "Cylinder", color)
{

}


Cylinder* Cylinder::operator-(const Shape3D& s) const
{
    Rectangle* newbaseShape = static_cast<Rectangle*>(baseShape)->operator -(*s.getBaseShape());
    Cylinder* newCylinder = new Cylinder(buildCylinder(newbaseShape),newbaseShape, getColor()->operator -(*s.getColor()));
    return newCylinder;
}



Cylinder* Cylinder::operator+(const Shape3D& s) const
{
    Rectangle* newbaseShape = static_cast<Rectangle*>(baseShape)->operator +(*s.getBaseShape());
    Cylinder* newCylinder = new Cylinder(buildCylinder(newbaseShape),newbaseShape, getColor()->operator +(*s.getColor()));
    return newCylinder;
}



Cylinder* Cylinder::operator*(const Shape3D& s) const
{
    Rectangle* newbaseShape = static_cast<Rectangle*>(baseShape)->operator *(*s.getBaseShape());
    Cylinder* newCylinder = new Cylinder(buildCylinder(newbaseShape),newbaseShape, getColor()->operator *(*s.getColor()));
    return newCylinder;
}



Cylinder* Cylinder::operator/(const Shape3D& s) const
{
    Rectangle* newbaseShape = static_cast<Rectangle*>(baseShape)->operator /(*s.getBaseShape());
    Cylinder* newCylinder = new Cylinder(buildCylinder(newbaseShape),newbaseShape, getColor()->operator /(*s.getColor()));
    return newCylinder;
}



double Cylinder::height() const
{
    //return volume()/(M_PI * static_cast<Rectangle*>(baseShape)->base());
    return static_cast<Rectangle*>(baseShape)->height();
}

double Cylinder::ray() const
{
    return static_cast<Rectangle*>(baseShape) ->base();
}

double Cylinder::lateralSurface() const
{
    Rectangle* tmp = static_cast<Rectangle*>(baseShape);
    return 2 * M_PI * tmp->base() * tmp->height();
}



double Cylinder::baseSurface() const
{
    Rectangle* tmp = static_cast<Rectangle*>(baseShape);
    return M_PI * pow(tmp->base(),2);
}



double Cylinder::volume() const
{
    Rectangle* tmp = static_cast<Rectangle*>(baseShape);
    return M_PI * pow(tmp->base(),2) * tmp->height();
}



double Cylinder::totalSurface() const
{
    return lateralSurface() + 2 * baseSurface();
}
